cpp/src/arrow/io/io-file-test.cc

```diff
@@ -614,6 +614,66 @@ TEST_F(TestMemoryMappedFile, WriteRead) {
   }
 }
 
+TEST_F(TestMemoryMappedFile, WriteAt) {
+  const int64_t buffer_size = 1024;
+  std::vector<uint8_t> buffer(buffer_size);
+  test::random_bytes(buffer_size, 0, buffer.data());
+
+  std::string path = "io-memory-map-write-read-test";
+  std::shared_ptr<MemoryMappedFile> result;
+  ASSERT_OK(InitMemoryMap(buffer_size, path, &result));
+
+  int64_t position;
+  ASSERT_OK(result->WriteAt(0, buffer.data(), buffer_size / 2));
+  ASSERT_OK(result->Tell(&position));
+  ASSERT_EQ(position, buffer_size / 2);
```

Comment:
Interesting. I wouldn't expect `WriteAt` to change the internal position at all. I'd rather remove those lines from the test (and perhaps `WriteAt` to remove the seeking, which hopefully won't break anything :-)).

Reply:
It doesn't break any tests, at least. But `FixedSizeBufferWriter` and `CudaMemory` implement `WriteAt` in such a way, that it moves the position. 

I've pushed the commit changing `WriteAt` behaviour for the `MemoryMappedFile`, but I'd rather revert it: I think a bit surprising, but consistent behaviour is preferable here.

Reply:
Ok, I guess we can fix that in another PR. But I'd rather not have the tests ascertain the current behaviour :-)

Reply:
Removed tests, kept behaviour for now.